Best-effort co-concretization for environments with explicit conflicts

becker33 opened the pull request:
Currently, environments can either be concretized fully together or fully separately. This works well for users who create environments for interoperable software and can use `concretization: together`. It does not allow environments with conflicting software to be concretized for maximal interoperability.

The primary use-case for this is facilities providing system software. Facilities provide multiple MPI implementations, but all software built against a given MPI ought to be interoperable.

This PR adds a `concretization` option `together_where_possible`. When this option is used, Spack will concretize specs in the environment separately, but will optimize for minimal differences in overlapping packages.

TODO:
- [x] tests
- [ ] documentation
- [x] improve optimization criterion for minimizing differences
- [x] minimize appearances of a given virtual

Stretch goal:
- [ ] dynamic optimization without writing to file

becker33 committed:
initial prototype of process spaces

no effort to minimize spaces, always use a single process space

becker33 committed:
remove unused node_compiler_set function

becker33 committed:
prioritization capabilities for minimizing duplicate nodes

becker33 committed:
remove incorrect TODOs

becker33 committed:
Add best effort coconcretization for environments

Involves maximizing reuse across process spaces in concretize.lp
Interface naming options may need to change

becker33 commented:
@alalazo this is the PR we were discussing earlier

becker33 committed:
flake

becker33 committed:
fix constraint msgs

becker33 committed:
fixup diff command

becker33 committed:
fixup diff tests

becker33 committed:
rename for clarity

becker33 committed:
per-package minimization of non-identical nodes

becker33 committed:
rename multi_root option for clarity

becker33 committed:
minimize duplicate virtual providers

becker33 committed:
add tests for best-effort coconcretization

alalazo commented on lib/spack/spack/cmd/solve.py:
```diff
@@ -105,7 +105,7 @@ def solve(parser, args):
     result = asp.solve(
         specs, dump=dump, models=models, timers=args.timers, stats=args.stats,
-        reuse=args.reuse,
+        reuse=args.reuse, multi_root=True
```
The solve command is cluttered because of the number of optimization criteria. A typical `spack solve` shows hundreds of lines of minimizations over possible dependencies - this seems a usability issue similar to lot of facts being dumped for errors.

becker33 replied:
Should be fixed in [9648b56](https://github.com/spack/spack/pull/27964/commits/9648b5652c3458658d566ca82f2140c4b10656c9)

alalazo replied:
I see weird output. Here's what `zlib` looks like:
![Screenshot from 2022-02-08 10-19-13](https://user-images.githubusercontent.com/4199709/152956472-4843d4a3-7678-4aeb-b13c-719161baee62.png)

and here's `hd5~mpi`:

![Screenshot from 2022-02-08 10-23-32](https://user-images.githubusercontent.com/4199709/152956921-1931c40e-f9f6-40db-966a-dd9774b01ac2.png)

alalazo replied:
Reading the code I see that now we show by default only the criteria that are non-zero i.e. sub-optimal. Is there an easy way to show output like before in non-debug mode and add:
```
number of non-identical *** nodes 
```
criteria only in debug mode?

alalazo commented on lib/spack/spack/concretize.py:
```diff
@@ -769,6 +770,10 @@ def make_concretization_repository(abstract_specs):
 
+    if kwargs.get('multi_root', False):
+        # This feature cannot be implemented in the old concretizer
+        raise Exception
```
Minor, but marking this line since we should raise an object and be a bit more specific.

becker33 replied:
Should be fixed by [62d7fc2](https://github.com/spack/spack/pull/27964/commits/62d7fc28b8bdf3c6ac4cf9d90fe168231a64a053)

alalazo commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -68,8 +68,9 @@ def compare_specs(a, b, to_string=False, color=None):
 
-    a_facts = set(t for t in setup.spec_clauses(a, body=True, expand_hashes=True))
-    b_facts = set(t for t in setup.spec_clauses(b, body=True, expand_hashes=True))
+    # psid index doesn't matter
```
Would this be true when we generalize concretization to allow multiple specs from the same package for a single root?

tgamblin replied:
also: say why in the comment.

becker33 replied:
Todd's concern should be fixed by [860694b](https://github.com/spack/spack/pull/27964/commits/860694b566badd3ae8258d0420fe480213204909)

Max's concern I think I need to wait for the PR that adds separable concretization of build deps to know how to handle this then.

alalazo commented on lib/spack/spack/schema/env.py:
```diff
@@ -150,7 +150,7 @@
                     'concretization': {
                         'type': 'string',
-                        'enum': ['together', 'separately'],
+                        'enum': ['together', 'separately', 'together_where_possible'],
```
Minor: wondering if we should come up with a single word or anyhow something shorter for this concretization mode.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -399,12 +414,19 @@ def specs(self):
+            # If everything has to fit in one process space, then only use
+            # process space 0
+            if not self.multi_root:
+                idx = 0
```
Can we use another name for the process index, so that we keep the current iteration index? 

EDIT: Maybe it would be good to construct the `(psid, root_spec)` key tuple in `__init__` and reuse that afterwards to fix the association in one place?

becker33 replied:
Fixed in [62e9b5b](https://github.com/spack/spack/pull/27964/commits/62e9b5b3407150d1b673f92ec28d5fe691eddf7d)

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -504,7 +538,9 @@ def fact(self, head, assumption=False):
     def solve(
             self, solver_setup, specs, dump=None, nmodels=0,
             timers=False, stats=False, tests=False, reuse=False,
+            multi_root=False
```
Not a fault of this PR, and can be done separately - but we should add a docstring to this method as it is gaining new arguments at each new feature.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -504,7 +538,9 @@ def fact(self, head, assumption=False):
             timers=False, stats=False, tests=False, reuse=False,
+            multi_root=False
     ):
+        # Spec_list must always be treated in order so PSIDs match up
```
Not clear what this comment means. Can we avoid pitfalls like that if we construct the matching in one place and then reuse it?

becker33 replied:
Fixed in [fe4c7ed](https://github.com/spack/spack/pull/27964/commits/fe4c7ed7689dc54a049d1e774505155562078683)

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -518,9 +554,37 @@ def solve(
             self.backend = backend
             solver_setup.setup(self, specs, tests=tests, reuse=reuse)
+
+            strings = []
```
This might use a more specific name e.g. `dynamic_minimization_rules`

becker33 replied:
Fixed in [22f2616](https://github.com/spack/spack/pull/27964/commits/22f2616f4ccee3b54aae6cdf5dc4a8c030a7e153)

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -518,9 +554,37 @@ def solve(
+                    template_virtual_opt.format(priority, virt),
+                    template_virtual_null.format(priority),
+                    template_virtual_present.format(priority, virt)
+                ])
```
I am a bit concerned over these priorities being arbitrary **and** separated. For `hdf5` we have numbers as high as 761 for the optimization priorities - I assume that we have 400-500 different optimization criteria all at different priorities. To give context of what I mean:
```console
$ spack solve --show=asp hdf5
...
#minimize { 0@758 : #true}.
#minimize { 1@758,PSID1,PSID2 : virtual_not_equal(PSID1, PSID2, "unwind")}.
opt_criterion(759, "number of non-identical uuid providers").
#minimize { 0@759 : #true}.
#minimize { 1@759,PSID1,PSID2 : virtual_not_equal(PSID1, PSID2, "uuid")}.
opt_criterion(760, "number of non-identical xxd providers").
#minimize { 0@760 : #true}.
#minimize { 1@760,PSID1,PSID2 : virtual_not_equal(PSID1, PSID2, "xxd")}.
opt_criterion(761, "number of non-identical yacc providers").
#minimize { 0@761 : #true}.
#minimize { 1@761,PSID1,PSID2 : virtual_not_equal(PSID1, PSID2, "yacc")}.
...
```
So far I tried cases where this doesn't matter, but it's absolutely not clear to me why we should optimize for having duplicate `xz` nodes at lower priority than `zlib` or vice-versa. What would be the drawback of having a single minimization objective for the sum of all these nodes?

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -669,39 +735,45 @@ def key_fn(version):
             self.gen.fact(fn.deprecated_version(pkg.name, v))
 
-    def spec_versions(self, spec):
+    def spec_versions(self, idx, spec):
```
In general I think we need to improve docstrings now that the arguments are becoming more involved conceptually. As far as I understand here `idx` is not arbitrary, but tied to the process ID / root ID? If there is a relation between these two arguments that needs to be respected, can't we package a spec and an idx in an object and send them around together?

becker33 replied:
The same spec could appear as part of multiple psids, since it could be a dependency of multiple roots.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -869,11 +941,14 @@ def condition(self, required_spec, imposed_spec=None, name=None):
         self.gen.fact(fn.condition(condition_id))
 
         # requirements trigger the condition
+        # index does not matter, its simply a required argument to generate clauses
```
I think this is not ideal for readability. We have an API with a mandated argument that is not needed in all contexts, and the caller has to strip it explicitly. I am wondering if we can insert a thin middle layer to avoid this, and pass specs which have or have not an `idx` associated with them.

becker33 replied:
Should be fixed by [944b3d4](https://github.com/spack/spack/pull/27964/commits/944b3d49c18d5959db807ad9b430796c7e816ddf)

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -882,14 +957,19 @@ def condition(self, required_spec, imposed_spec=None, name=None):
     def impose(self, condition_id, imposed_spec, node=True, name=None, body=False):
+        # idx does not matter, just needed to generate clauses
+        # it would be much more complicated to generate clauses with/without
+        # process space ids
```
This is another place where we have to strip `idx`. There's probably an abstraction missing in the current implementation :thinking:

becker33 replied:
Should be fixed by [944b3d4](https://github.com/spack/spack/pull/27964/commits/944b3d49c18d5959db807ad9b430796c7e816ddf)

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1112,7 +1200,7 @@ class Head(object):
             node_os = fn.node_os_set
             node_target = fn.node_target_set
             variant_value = fn.variant_set
-            node_compiler = fn.node_compiler_set
```
This issue can be fixed in a separate PR, but nice catch!

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1381,13 +1468,18 @@ def target_defaults(self, specs):
+            # If we aren't allowing node splits, then we only need a single
+            # process space, use 0
+            if not self.gen.multi_root:
+                idx = 0
```
Here we can probably reuse a tuple computed in advance (to be clear, I am more concerned about consistency when this code will be changed later than performance). If I understand correctly we have multiple places where we iterate over the same list of specs and recompute the idx. This is fragile with respect to changes to the list that is passed around.

becker33 replied:
Fixed in [fe4c7ed](https://github.com/spack/spack/pull/27964/commits/fe4c7ed7689dc54a049d1e774505155562078683)

alalazo reviewed (commented):
I started reviewing the PR, and left a few questions. At a high level:
- The `spack solve output` is very verbose with hundreds of optimization criteria
- The association of a process ID and a spec is maybe too loose in the code, and needs to be consolidated in some way.
- Having separate optimization criteria for each package poses questions on how to select priorities

I tested the solve time for a few specs and the results are great, i.e. we don't have noticeable performance losses. Solve time for PR 27694 (in seconds):

| | **hdf5** | **cp2k** |**trilinos** |
|-|----------|----------|-------------|
| *setup* | 3.6892 | 5.6508 | 4.7846 |
| *ground* | 1.3216| 2.2394 | 1.9163 |
| *solve* | 2.4108 | 4.1936 | 2.8291 |
| **total** | 7.4647 | 12.1481 | 9.5782 |

Solve time for `develop` (in seconds):

| | **hdf5** | **cp2k** |**trilinos** |
|-|----------|----------|-------------|
| *setup* | 3.6545 | 5.4729 | 4.6592 |
| *ground* | 1.0601| 1.6933 | 1.5260 |
| *solve* | 2.1859 | 3.7101 | 3.4006 |
| **total** | 6.9415 | 11.0188 | 9.6327 |

This comparison points out that the new encoding doesn't add much overhead when we fall-back to the previous case of having a single process id.

Tests run on:

* **Spack:** 0.17.1-710-afa4221598
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

alalazo commented:
Another data point by test-driving the PR with this simple environment:
```yaml
spack:
  specs:
  - hdf5@1.13.0 ^zlib@1.2.8
  - hdf5@1.8.22
  concretization: together_where_possible
```
I obtain:
```console
$ time spack -e . concretize -f
...
real	0m20,139s
user	0m19,811s
sys	0m0,309s
```

Compared to concretizing separately on `develop`:
```console
$ time spack -e . concretize -f
==> Starting concretization pool with 2 processes
...

real	0m8,780s
user	0m16,109s
sys	0m0,335s
```

There is a 20% increase in user time when we have to unify zlib over 2 specs. Using a slightly more complex environment:
```yaml
spack:
  specs:
  - hdf5@1.13.0 ^zlib@1.2.8
  - hdf5@1.8.21 ^zlib@1.2.11
  - hdf5@1.8.22
  concretization: together_where_possible
```
I have:
```console
$ time spack -e . concretize -f
...
real	1m2,460s
user	1m1,971s
sys	0m0,461s
```

Compared to the separate concretization:
```console
$ time spack -e . concretize -f
...
real	0m8,966s
user	0m24,255s
sys	0m0,514s
```
which is a 150% increase in user time. Furthermore, `hdf5@1.8.22` is concretized using `zlib@1.2.8` which I think is unexpected / sub-optimal, see screenshot:

![2022-01-04_11-16](https://user-images.githubusercontent.com/4199709/148044729-82ce03b5-a311-449a-8013-7ecd90354376.png)

alalazo commented:
Using this environment:
```yaml
spack:
  specs:
  - hdf5@1.13.0 ^zlib@1.2.8 ^xz@5.2.4
  - hdf5@1.8.21 ^zlib@1.2.11
  - hdf5@1.8.22
  - xz@5.2.5
  concretization: together_where_possible
```
Gives very weird results where the second and third root specs have the `mpi` variant flipped to `~mpi`. 
![2022-01-04_11-35](https://user-images.githubusercontent.com/4199709/148046490-7b3226f9-25bc-44a1-a18a-aaa147bd5763.png)


I'm attaching the lockfile in case it is of interest.

[spack.lock.txt](https://github.com/spack/spack/files/7806968/spack.lock.txt)

becker33 committed:
do not overcount differences

becker33 committed:
asp.py: break version ties from specs by version, not order of appearance

becker33 committed:
show only non-zero opt criteria without debug mode

becker33 commented:
@alalazo I have addressed the two behavior bugs you showed. Still working on addressing other review comments.

becker33 committed:
proper error handling

becker33 committed:
flake

becker33 committed:
use psid for process space identifier in all instances

becker33 committed:
remove psid from preferred_target_weight

becker33 committed:
enumerate psids once and use canonical source in all cases

becker33 committed:
rename variable for clarity

becker33 committed:
better interface for ignoring psid in abstract clauses

becker33 committed:
no psids for `spack diff`

becker33 committed:
fix representatives for virtuals

becker33 committed:
flake

becker33 committed:
flake

becker33 committed:
revert change to diff tests

becker33 committed:
fix virtual_not_equal

becker33 committed:
test concretization preferences with coconcretization

alalazo commented:
I tried again the same quick performance test. Here's the solve time for `develop` at c1b51d6e997667120f31b27efbf9d5c99e540566 for reference (in seconds):

| | **hdf5 (ahqceuw)** | **cp2k (rhkmujf)** |**trilinos (7k3phbh)** |
|-|----------|----------|-------------|
| *setup* | 3.8275 | 5.6943 | 4.8123 |
| *ground* | 1.2049| 1.7854 | 1.6058 |
| *solve* | 2.2181 | 4.6457 | 3.2920 |
| **total** | 7.2918 | 12.1834 | 9.7555 |


and here are the results for the PR at 78bb86e662e3e8848d37a169200d89805383f1d6 (merge commit, in seconds):

| | **hdf5 (ahqceuw)** | **cp2k (rhkmujf)** |**trilinos (7k3phbh)** |
|-|----------|----------|-------------|
| *setup* | 3.8857 | 5.9796 | 5.0564 |
| *ground* | 1.4768| 2.2562 | 2.0109 |
| *solve* | 2.4698 | 4.2704 | 3.3387 |
| **total** | 7.8765 (+8%)| 12.5708 (+3%) | 10.4562 (+7%) |
 
So overall just a small increase after the latest fixes, which I think is acceptable.

alalazo commented:
I confirm that this environment:
```yaml
spack:
  specs:
  - hdf5@1.13.0 ^zlib@1.2.8 ^xz@5.2.4
  - hdf5@1.8.21 ^zlib@1.2.11
  - hdf5@1.8.22
  - xz@5.2.5
  concretization: together_where_possible
```
now concretize as expected. I am attaching the [spack.lock](https://github.com/spack/spack/files/8022731/spack.lock.txt) file as a reference. 

It is concerning though that the time to reach a solution for this environment is:
```
real	2m34,696s
user	2m34,142s
sys	0m0,520s
```
if compared to separate concretization:
```
real	0m10,294s
user	0m27,867s
sys	0m0,571s
```
The concerning part is that it seems there's a great increase moving from a single spec to be unified to 2 specs. Now the user time is 154 sec. vs. 27 sec. so more than 5x. @becker33 I wonder if unifying everything in a single optimization criteria can mitigate this increase, since the search space is collapsed into a single additional dimension. I have no idea if this is just a wrong intuition or if it can be supported by any work I don't know about yet.

alalazo commented:
Tried to construct a more complicated environment, where the new optimization criteria are more exercised:
```yaml
spack:
  specs:
  - hdf5@1.13.0 ^zlib@1.2.8 ^xz@5.2.4 ^autoconf@2.71
  - hdf5@1.8.21 ^zlib@1.2.11 ^hwloc@2.6.0
  - hdf5@1.8.22
  - autoconf@2.70
  - xz@5.2.5
  concretization: together_where_possible
```
The overall time to solution doesn't increase much with respect to the simpler environment used before:
```
real	2m54,811s
user	2m54,297s
sys	0m0,482s
```
For reference, if we concretize separately:
```
real	0m10,044s
user	0m28,316s
sys	0m0,573s
```

Trying to concretize the e4s pipeline (as representative of a production environment) on my machine gives the following result on develop:
```
real	4m21,679s
user	30m46,242s
sys	0m6,934s
```
when trying `together_where_possible`:
```
$ time spack -e . concretize -f
Killed

real	5m31,647s
user	5m20,116s
sys	0m11,406s
```
as it literally hogs memory (in 5 mins. it ramped up to exceed the 32GB of RAM my machine has). I guess we need to keep an eye on memory too :sweat:

alalazo commented on lib/spack/spack/test/concretize.py:
```diff
@@ -1375,3 +1375,49 @@ def test_sticky_variant_in_package(self):
+        (['mpi', 'mpich'], 1),
+    ])
+    def test_best_effort_coconcretize(mock_packages, specs, expected):
+        import spack.solver.asp
+        specs = [spack.spec.Spec(s) for s in specs]
+        result = spack.solver.asp.solve(specs, reuse=False, multi_root=True)
```
Shouldn't we skip this test when using the original concretizer?

alalazo commented on lib/spack/spack/test/concretize.py:
```diff
@@ -1375,3 +1375,49 @@ def test_sticky_variant_in_package(self):
+        """
+        Test that package preferences are being respected during coconcretization.
+        """
+        import spack.solver.asp
+
+        specs = [spack.spec.Spec(s) for s in specs]
+        result = spack.solver.asp.solve(specs, reuse=False, multi_root=True)
```
Same question on skipping the test for `concretizer:original`

alalazo reviewed (changes requested):
I am halfway through the review. So far I have only [one minor question](https://github.com/spack/spack/pull/27964/files#r801424777) on the output of `spack solve` which is probably relevant only for developers. 

I tried to test drive the PR and this time the results have been always correct, so :100: for the fixes on behavior.

I am a bit concerned by resources that this mode of concretization needs. So far I have been unable to concretize `e4s` (used as a representative of a production environment) on my machine due to memory requests.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -98,9 +99,38 @@ def getter(node):
+
+# Class for negating version ranges
+# This allows us to sort ascending by inverse version order
+class _InvertedVersion(object):
+    def __init__(self, version):
+        self.version = version
+
+    def __eq__(self, other):
+        return self.version == other.version
+
+    def __lt__(self, other):
+        return self.version > other.version
+
+    def __gt__(self, other):
+        return self.version < other.version
+
+    def __ne__(self, other):
+        return self.version != other.version
+
+    def __le__(self, other):
+        return self.version >= other.version
+
+    def __ge__(self, other):
+        return self.version <= other.version
```
Can we use this instead:
```suggestion
@functools.total_ordering
class _InvertedVersion(object):
    def __init__(self, version):
        self.version = version

    def __eq__(self, other):
        return self.version == other.version

    def __lt__(self, other):
        return self.version > other.version
```
? Or are there performance issues with the decorator?

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -98,9 +99,38 @@ def getter(node):
 # Below numbers are used to map names of criteria to the order
 # they appear in the solution. See concretize.lp
 
+#: High fixed priority offset for criteria that supersede all build criteria
```
It's probably worthwhile to add a comment explaining how the space of criteria is reserved at a high level (feel free to reword anything you like):
```python
# The space of possible priorities for optimization targets 
# is partitioned in the following ranges:
#
# [0-100) Optimization criteria for software being reused
# [100-200) Fixed criteria that are higher priority than reuse, but lower than build
# [200-300) Optimization criteria for software being built
# [300-inf) High-priority fixed criteria
#
# Each optimization target is a minimization with optimal value 0.
```

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -115,6 +145,7 @@ def build_criteria_names(costs, tuples):
     priorities_names = []
 
     num_fixed = 0
+    num_high_fixed = 0
```
We should probably rename `num_fixed` for better clarity, but I have no good suggestion at the moment :slightly_frowning_face:

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -201,11 +241,17 @@ def _id(thing):
+# To avoid duplicate code paths for spec_clauses calls that do/don't require psid
+# we have the no_psid object. Any AspFunction created with psid set to `no_psid` will
+# strip that argument from its args.
+no_psid = object()
```
:+1:

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -528,9 +592,10 @@ def fact(self, head, assumption=False):
     def solve(
-            self, solver_setup, specs, dump=None, nmodels=0,
-            timers=False, stats=False, tests=False, reuse=False,
+            self, solver_setup, specs_by_psid, dump=None, nmodels=0,
```
Not an issue due to this PR, but it seems to me that the `dump` argument is unused in this method (it's used higher in the call stack). I think we can safely remove it.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2049,9 +2171,17 @@ def solve(specs, dump=(), models=0, timers=False, stats=False, tests=False,
                 continue
             spack.spec.Spec.ensure_valid_variants(s)
 
+    # Associate specs with psids so we don't have to worry about order later
+    if multi_root:
+        # each spec gets its own psid
+        specs_by_psid = list(enumerate(specs))
+    else:
+        # all specs are given psid 0
+        specs_by_psid = list(zip_longest([0], specs, fillvalue=0))
+
```
:+1:

alalazo reviewed (commented):
Left a few other minor comments.

becker33 committed:
solve cmd: add option for multi-root vs single-root

becker33 committed:
concretize.lp: sum multi-root badness across packages

becker33 committed:
remove unnecessary derivation for representative_virtual

becker33 committed:
show all opt criteria now that there are fewer

becker33 committed:
cmd/solve: fix handling of dump options

node: dump=output still does not work

alalazo committed:
Reduce the number of psids by adding a level of indirection

This commit introduce the "literal" atom, which comes with
a few different "arities". The unary "literal" contains an
integer that id the ID of a spec literal. Other "literals"
contain information on the requests made by literal ID. For
instance zlib@1.2.11 generates the following facts:

literal(0,"root","zlib").
literal(0,"node","zlib").
literal(0,"node_version_satisfies","zlib","1.2.11").

PSIDs are instead generated by the solver and minimized.

This should help with treating large environments, since
resource requests from clingo grow at least linearly with
the number of psids.

alalazo committed:
Removed psid argument from SpackSolverSetup methods

alalazo committed:
Fix typo in test method

alalazo committed:
Remove fine grain selection of packages

This PR removes the clauses that are deducing which
packages are equal across PSIDs and the minimizations
associated with them.

The reason for the removal is that those clauses are
growing quadratically in complexity.

This has been substituted by minimizing the number of
PSIDs and by penalizing the use of providers (i.e. number
of incoming edges) rather than just te selection of
providers (nodes).

alalazo committed:
Update command completion

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1004,21 +958,15 @@ equality_attr("external_spec_selected").
 opt_criterion(100, "number of packages to build (vs. reuse)").
 #minimize { 0@100: #true }.
-#minimize { 1@100,Package,PSID : build_representative(PSID, Package), optimize_for_reuse() }.
+#minimize { 1@100,Package : build(PSID, Package), optimize_for_reuse() }.
```
Are you sure this shouldn't be just the representative? Once we're building something, we shouldn't bias against reusing it in another spec.

alalazo replied:
No, not really sure. I think you're right it should be `build_representative` because at best here I am double counting the same spec twice if it comes from different PSIDs. This means though that we can't get rid of the rules that are growing quadratically with PSIDs in grounding.

alalazo replied:
Also, leaving `PSID` out in the lhs is a typo :slightly_frowning_face:

alalazo replied:
@becker33 @tgamblin Another data point: concretizing [the pipelines](https://gitlab.spack.io/spack/spack/-/pipelines/98600) took an insane amount of time (and e4s didn't concretize in the 6hrs limit) I think the main issue might be we are introducing a lot of symmetry with PSIDs. Think of this simplified case of allowing clingo to place specs in `M` different PSIDs. Assume `N` of the specs can be moved around from a PSID to the other because they don't interfere with each other. That gives `M^N` possible optimal combinations.

Of course this is just rough intuition, since the reality of a solve is much more dynamic, but it would be a nice thing to try to introduce some artificial symmetry breaking rule in the PSID space and see if we benefit from it.

tgamblin commented:
Superseded by #28941
